treeherder/log_parser/tasks.py

```diff
@@ -50,20 +32,39 @@ def parse_logs(job_id, job_log_ids, priority):
         "builds-4h": parse_unstructured_log
     }
 
+    # We don't want to stop parsing logs for most Exceptions however we still
+    # need to know one occurred so we can skip further steps.
+    a_parser_failed = False
     completed_names = set()
-    exceptions = []
     for job_log in job_logs:
+        newrelic.agent.add_custom_parameter("job_log_%s_url" % job_log.name, job_log.url)
+        logger.debug("parser_task for %s", job_log.id)
+
+        # Don't parse lobs which have already been parsed.
+        if job_log.status == JobLog.PARSED:
+            logger.info("%s log already parsed", job_log.id)
+            continue
+
         parser = parser_tasks.get(job_log.name)
-        if parser:
-            try:
-                parser(job_log)
-            except Exception as e:
-                exceptions.append(e)
-            else:
-                completed_names.add(job_log.name)
+        if not parser:
+            continue
+
+        try:
+            parser(job_log)
+        except Exception as e:
+            if isinstance(e, SoftTimeLimitExceeded):
+                raise
+
+            a_parser_failed = True
+
+            # track the exception on NewRelic but don't stop parsing future
+            # log lines.
+            newrelic.agent.record_exception()
+        else:
+            completed_names.add(job_log.name)
 
-    if exceptions:
-        raise exceptions[0]
+    if a_parser_failed:
+        return
```

Comment:
This changes the behaviour significantly, since now failed jobs of an eligible exception type won't be retried (`@retryable_task`).

Perhaps the only way to fix bug 1483267 without this becoming overly complex, is to instead to decide that we'll always parse say the raw text log first, and if that fails, don't bother with the rest (and since the job will be retried, if it was a transient issue it won't matter much anyway). That way we don't need any of this custom handling, and there are much fewer states we can end up in.

Reply:
Expanding on the last point slightly - I think the possible failure modes of the log parser, are:
1. some transient connectivity/queue.taskcluster.net/S3/Treeherder DB/... issue that would affect all log types, that will likely recover after retrying
2. a generic bug in the log parser that will never succeed after retrying, that affects all log types (eg bad code deploy)
3. a log-file-specific Exception that will never succeed after retrying (eg not handling astral unicode properly and a particular job log including such characters)

If we made the simplifications above, then there is no difference for (1) or (2), since all logs would have been affected anyway.

For (3), yes later logs would never get parsed, however:
* we can parse the most important log first (likely the raw text log) to minimise impact
* any such cases are real bugs we should fix ASAP, so shouldn't be common in practice

Reply:
I completely missed the retryable decorator, sorry!